src/Microsoft.ML.Data/Commands/DefaultColumnNames.cs

```diff
@@ -4,7 +4,13 @@
 
 namespace Microsoft.ML.Data
 {
-    public static class DefaultColumnNames
+    /// <summary>
+    /// A set of string literals intended to be "canonical" names for column names intended for particular purpose.
+    /// While not part of the public API surface, its primary purpose is intended to be used in such a way as to encourage
+    /// uniformity on the public API surface, wherever it is judged where columns with default names should be consumed.
+    /// </summary>
+    [BestFriend]
+    internal static class DefaultColumnNames
```

Comment:
While helpful for *enabling* a uniform public surface, this is of course not the same as it needing to be in the public surface.

Reply:
(I agree we should internalize this - just raising as a general question in this area)

How do we imagine users to find out about these "magic" strings? Currently, one way is the default values in the MLContext APIs. But I wonder if that is enough. For example, when making a Prediction "output" class, how do users know to name/attribute their properties as `PredictedLabel`, `Probability`, `Score`?  Just through documentation?

---
In reply to: [262298366](https://github.com/dotnet/machinelearning/pull/2842#discussion_r262298366) [](ancestors = 262298366)

Reply:
Yes, unfortunately I think both helpful displays in the debugger (for live discovery) and documentation (for offline discovery) are the only real answer we have right now -- both seem insufficient or at least suboptimal. This is a problem I think in most places we've chosen to believe that producing "default" columns names is more helpful than requiring the user to specify them. I am unsure how to address that problem. Something *resembling* the static typed API will I think be a solution to that problem, since the columns will be typed and addressable and right there.

Reply:
I have seen quite a few instances where users use `DefaultColumnNames`. I don't understand why we want to hide it. We will always have default column names, why not allowing users to access them instead of hard coding them in their applications?

searching the samples repo shows there are quite a few instances where people use `DefaultColumnNames`.
[https://github.com/dotnet/machinelearning-samples/search?q=defaultcolumnnames&unscoped_q=defaultcolumnnames](https://github.com/dotnet/machinelearning-samples/search?q=defaultcolumnnames&unscoped_q=defaultcolumnnames)

or this other user provided code:
[https://github.com/dotnet/machinelearning/issues/2486#issuecomment-462248105](https://github.com/dotnet/machinelearning/issues/2486#issuecomment-462248105)


---
In reply to: [262302369](https://github.com/dotnet/machinelearning/pull/2842#discussion_r262302369) [](ancestors = 262302369)

Reply:
In those places, what would have happened in those samples if instead of `DefaultColumnNames.Features` we instead wrote `"Features"`? I understand the appeal inside the tool of having a consistent "location" for them. For user code, I am less certain.

I am very uncomfortable with having as part of our public surface a class that is merely a collection of "magic strings" that we've used internally to denote certain concepts. That's not great API design.

The samples can and should change. To use this class of magic constants was, frankly, a bad move by the sample author, since it was totally needless.